src/3d/shaders/postprocess.frag

```diff
@@ -0,0 +1,141 @@
+#version 150 core
+
+uniform sampler2D colorTexture;
+uniform sampler2D depthTexture;
+//uniform sampler2DShadow shadowTexture;
+uniform sampler2D shadowTexture;
+
+// light camera uniforms
+uniform mat4 viewMatrix;
+uniform mat4 projectionMatrix;
+uniform float lightFarPlane;
+uniform float lightNearPlane;
+
+uniform float shadowMinX;
+uniform float shadowMaxX;
+uniform float shadowMinZ;
+uniform float shadowMaxZ;
+
+uniform vec3 lightPosition;
+uniform vec3 lightDirection;
+
+// view camera uniforms
+uniform mat4 cameraView;
+uniform mat4 cameraProj;
+uniform float farPlane;
+uniform float nearPlane;
+
+uniform int renderShadows;
+uniform float shadowBias;
+
+in vec2 texCoord;
+
+out vec4 fragColor;
+
+vec3 WorldPosFromDepth(float depth) {
+    float z = depth * 2.0 - 1.0;
+    mat4 projMatrixInv = inverse(cameraProj);
+    mat4 viewMatrixInv = inverse(cameraView);
+
+    vec4 clipSpacePosition = vec4(texCoord * 2.0 - 1.0, z, 1.0);
+    vec4 viewSpacePosition = projMatrixInv * clipSpacePosition;
+
+    viewSpacePosition /= viewSpacePosition.w;
+    vec4 worldSpacePosition =  viewMatrixInv * viewSpacePosition;
+    worldSpacePosition /= worldSpacePosition.w;
+
+    return worldSpacePosition.xyz;
+}
+
+float LinearizeDepth(float depth)
+{
+  float nearPlane = lightNearPlane;
+  float farPlane = lightFarPlane;
+    float z = depth * 2.0 - 1.0; // back to NDC
+    float d = (2.0 * nearPlane * farPlane) / (farPlane + nearPlane - z * (farPlane - nearPlane));
+    return d / farPlane;
+}
+
+float CalcShadowFactor(vec4 LightSpacePos)
+{
+  vec2 texelSize = 1.0 / textureSize(shadowTexture, 0);
+  vec3 ProjCoords = LightSpacePos.xyz / LightSpacePos.w;
+  vec2 UVCoords;
+  UVCoords.x = 0.5 * ProjCoords.x + 0.5;
+  UVCoords.y = 0.5 * ProjCoords.y + 0.5;
+  float z = 0.5 * ProjCoords.z + 0.5;
+
+  // percentage close filtering of the shadow map
+  float shadow = 0.0;
+  int k = 1;
+  for(int x = -k; x <= k; ++x)
+  {
+    for(int y = -k; y <= k; ++y)
+    {
+      float pcfDepth = texture(shadowTexture, UVCoords + vec2(x, y) * texelSize).r;
+      shadow += z - shadowBias >= pcfDepth ? 0.5 : 1.0;
```

Comment:
0.5 is the visibility of a pixel when it's in shadow, 1.0 is for normally shaded pixels.
Maybe I should add shadow darkness parameter instead of 0.5 ?